keps/168-2-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,292 @@
+
+The API does not allow for the modification of any objects.
+
+Regular users will have access only to the LocalQueues they are assigned to. However, they will be able to fetch information about the global position of a workload in a ClusterQueue, without any details about workloads in different LocalQueues. 
+
+Administrators will have access to all the data at the ClusterQueue level. They will be able to view all the workloads, no matter the LocalQuees the workloads are assigned to. 
+
+### URIs:
+
+We introduce a new API group ```pending-workloads.kueue.x-k8s.io``` that aggregates the following endpoints:
+
+#### List all pending workloads in ClusterQueue
+
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME
+```
+
+#### List all pending workloads in LocalQueue
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/namespaces/LQ_NAMESPACE/localqueues/LQ_NAME
+```
+
+#### Fetch information about single Workload
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/namespaces/WL_NAMESPACE/workloads/WL_NAME
+```
+
+### API Objects:
+
+```
+// User-facing representation of a PendingWorkload in both LocalQueues and ClusterQueue that summarizes neccessary information from the admission order perspective
+type PendingWorkloadSummary struct {
+	TypeMeta TypeMeta
+	ObjectMeta ObjectMeta
+
+	LocalQueueNamestring
+	PositionInClusterQueue int32
+	PositionInLocalQueue int32
+	Priority int32
+}
+
+type PendingWorkloadsSummaryList struct {
+	TypeMeta TypeMeta
+	ListMeta ListMeta
+
+	Items []PendingWorkloadSummaryInClusterQueue
+}
+```
+
+### KueueManager changes
+
+We propose changing KueueManager's cache, so it contains more information about pending workloads. This allows us to expose that information. We'll need cache:
+- workload's priority,
+- workload's LocalQueue name,
+- workload's creation timestamp,
+- relevant metadata.
+
+Workload's position in a ClusterQueue and a LocalQueue will be computed on request.
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Overview
+
+Our main focus is integration tests, as most of the added code is responsible for integrating with the Kueue and RBAC roles.
+
+#### Unit Tests
+
+We plan on adding unit tests that cover getting a list of pending workloads at the KueueManager level.
+
+- `pkg/visibility`: `30 Oct 2023` - `0%`
+
+#### Integration tests
+
+The integration tests should check if:
+- Pending workloads are returned according to their actual status without delay. This includes adding and removing/admitting new workloads with various priorities,
+- Adding workloads to one LocalQueues results in position changes for workloads submitted to other LocalQueues,
+- Data is consistent across all LocalQueues
+- User with only LocalQueue visibility can view only appropriate LocalQueues
+
```

Comment:
- User with only LocalQueue visibility cannot access the list of pending workloads for ClusterQueue